doni/driver/worker/blazar.py

```diff
@@ -1,20 +1,270 @@
+        auth_conf.register_auth_opts(conf, self.opt_group, service_type="reservation")
+
+    def list_opts(self):
+        return auth_conf.add_auth_opts(self.opts, service_type="reservation")
+
     def process(
         self,
         context: "RequestContext",
         hardware: "Hardware",
         availability_windows: "list[AvailabilityWindow]" = None,
-        state_details: "dict" = None,
+        state_details: "dict" = {},
     ) -> "WorkerResult.Base":
-        pass
+        """Main loop for Blazar sync worker.
+
+        This method ensures that an up-to-date blazar host object exists for
+        each physical host in Doni's DB.
+
+        "name" must match the name used by nova to identify the node. In our case
+        it is the hardware uuid, as that is what ironic is passing to nova.
+        blazar uses nova.get_servers_per_host to check if there is an existing
+        server with that name.
+        """
+
+        def _search_hosts_for_uuid(hw_uuid) -> dict:
+            host_list_response = _call_blazar(
+                context,
+                f"/os-hosts",
+                method="get",
+                json={},
+                allowed_status_codes=[200],
+            )
+            host_list = host_list_response.get("hosts")
+            matching_host = next(
+                (host for host in host_list if host.get("name") == hw_uuid),
+                None,
+            )
+            return matching_host
+
+        # If we know the host_id, then update that host.
+        # If we don't then attempt to create it
+        # we'll always "touch" the host, because we can't tell if this was a host
+        # or a lease update request yet.
+        result = {}
+
+        host_id = state_details.get("blazar_host_id")
+        if host_id:
+            # Always try to update the host in blazar. We could add a precondition
+            # header based on e.g. timestamp if needed.
+            try:
+                update = _call_blazar(
+                    context,
+                    f"/os-hosts/{host_id}",
+                    method="put",
+                    json=_blazar_host_requst_body(hardware),
+                    allowed_status_codes=[200],
+                )
+                result["blazar_host_id"] = update.get("id")
+                result["host_updated_at"] = update.get("updated_at")
+            except BlazarAPIError as exc:
+                # TODO what error code does blazar return if the host has a lease already?
+                if exc.code == 404:
+                    host = _search_hosts_for_uuid(hardware.uuid)
+                    if host:
+                        # update stored host_id with match, and retry after defer
+                        result["blazar_host_id"] = host.get("id")
+                    else:
+                        # remove invalid stored host_id and retry after defer
+                        result["blazar_host_id"] = None
+                    return WorkerResult.Defer(result)
```

Comment:
Cool, I like the use of defer here, keeps the function pretty clean if you don't have to re-process the host ID in the same run.